### PR TITLE
Architecture notes

### PR DESCRIPTION
**Pull request by mathemancer**
This adds architecture notes regarding the Models, Users and Permissions, and a quick overview.

[Rendered index](https://github.com/mathesar-foundation/mathesar-wiki/blob/architecture_notes/docs/engineering/architecture/index.md) | [Rendered models](https://github.com/mathesar-foundation/mathesar-wiki/blob/architecture_notes/docs/engineering/architecture/models.md) | [Rendered permissions](https://github.com/mathesar-foundation/mathesar-wiki/blob/architecture_notes/docs/engineering/architecture/permissions.md)

### CODE REVIEW

**Review by seancolsen — COMMENTED** [review]
Very cool. Thanks for sending this, @mathemancer!

## Comments on models.md

- I'm a little confused about the `Database`, `DatabaseServerCredential`, and `UserDatabaseMap` models. Do you think you could spell these out more explicitly using markdown tables that show the name and type of each column? In particular, the `UserDatabaseMap` model seems to have a `database_id` field _and_ and `database` field which I find weird. Perhaps this is a typo. Or maybe I'm not understanding. Seeing all columns and their types would help me. Also, the `Database` model, on its own seems strange. For it to have only an OID and name without any other fields seems like it's kind of weird and detached. I guess I would expect these fields to be present directly in `UserDatabaseMap` instead of in a separate table. I think I'm missing something.

- > This stores the template defining what should be shown in a referencing fkey column for this table

    Within this wiki page, I'd suggest changing the above sentence to:

    > This stores the record summary template for a table.

    "Record summary template" is the term that we use within the UI and I think it would be nice to use that consistently within our internal docs too.

    Additionally, if you do find yourself making changes to this model, then I'd suggest trying to adopt that "record summary template" terminology in the codebase too.

    "Record summary" is the text that gets rendered within the pill.

    "Record summary template" is the user-defined template that configures the fields to render.

    You also said:

    > This would be _much_ better as a `ma_catalog` table for efficiency reasons.

    Note that in my vision for [record summaries within worksheets](https://wiki.mathesar.org/engineering/specs/worksheets-technical-specs/#record-summaries), the record summary template would be stored within the worksheet definition. I think we ought to consider this idea more thoroughly before making any big changes to record summaries. Likewise I have been ruminating on some more radical changes to record summaries which would make them much more powerful. I want to chat about that too before we make any big changes. As currently implemented, record summaries have some annoying limitations in terms of functionality. And I understand they are also a culprit of poor performance.

- I'd be curious to see your proposal for what, specifically, the `TableMetadata` and `ColumnMetadata` models would look like. Would all fields be in one JSON blob? Or would each field be in a separate column? One reason I ask is that Mukesh and I had quite a bit of back and forth about updating column widths. He summarized it well in this [comment](https://github.com/mathesar-foundation/mathesar/issues/1421#issuecomment-1608443136). He wanted the API to accept a _full_ JSON blob of display options for a column, all at once. I wanted the API to accept a _partial_ JSON blob of _some_ display options and then to merge that data received from the API request with the data already saved. For some reason that proved to be tricky. I think this use-case is worth considering as we consider architectural changes.

## Comments on index.md

- Generally, this all looks great.

- At some point before implementation I'd like to discuss the list of RPC functions that we plan to implement. I want to assess them as a cohesive body. Here's an example of some weirdness we currently have: the `GET /api/db/v0/tables/` endpoint you mention returns column data. But the front end doesn't actually use that column data — it relies on the columns endpoint for that. If we can find ways to make the RPC functions more mutually exclusive in terms of purpose, I think that would help to simplify things. Personally, I would prefer us to use API call that returns the entire structure of a given schema — all tables, columns, constraints, indexes, etc. Pavish and I have debated this. Even if we did agree on moving in this direction, I don't think the beta would necessarily be the most opportune time to make a large change like that. But I think at the very least it's worth having a discussion to plan out the set of RPC endpoints so that we all agree they work well as a set, with consistent patterns.

## Comments on permissions.md

- I'm withholding review of this part until I have time to read Pavish's [Permissions doc](https://hackmd.io/@mathesar/Hkads67nT) as well.

File: docs/engineering/architecture/index.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+### Python `db` library
+
+This library should mostly serve to provide thin wrapper functions around the User DB layer functions. These functions should take parameters from requests (never request objects themselves) and engines, and then call underlying DB functions. They should then pass the results up towards the API
+
+### Web service
+
+This service should provide an JSON-RPC API for use by the front end. When an API function is called, the service should:
+
+- Grab an appropriate engine from the `DatabaseConnection` model by combining the user associated with the request with the `database`.
```
Comment (kgodey) [inline docs/engineering/architecture/index.md:115]: What `DatabaseConnection` model? Could you describe the new model structure somewhere and compare it to the current structure?

Comment (kgodey) [reply docs/engineering/architecture/index.md:115]: Never mind, found the models file, but can you link it somewhere from here?

Comment (kgodey) [reply docs/engineering/architecture/index.md:115]: Okay I found the models file but I am finding it very hard to follow with the ad-hoc notes, can we do some sort of comparison of old structure vs. new structure?

**Review by kgodey — COMMENTED** [review]
@mathemancer Here's my initial review, I'll review the permissions after I talk to Pavish about them tomorrow.

File: docs/engineering/architecture/index.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+### User database
+
+For each API call, there should be an identifiable DB function that performs all User Database operations needed to satisfy that call. For example,
+
+- Calling the function `get_tables(database, schema)` should result in a call to some function `get_tables(sch_oid oid)` on the `database`.
+
+To achieve this, we will install the following on the user database(s)
+
+- Some custom Mathesar types.
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/index.md:103]: Yes, will do.

File: docs/engineering/architecture/index.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+- Some tables holding metadata that needs to be kept near the tables.
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/index.md:104]: - This is mostly referring to the record summary template metadata. The current algorithm requires lots of back and forth between the table definition and the template definition, which (with our current architecture) is a serious performance drag.
- We will put these into a separate mathesar schema (`msar_cat` probably).

File: docs/engineering/architecture/index.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+- A set of functions that provide the bulk of Mathesar's back end logic and functionality.
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/index.md:105]: I'll flesh out this section in a new page, I think.

File: docs/engineering/architecture/index.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+## Bird's Eye View
+
+The main goals of this architectural redesign are to improve the speed and reduce the complexity of the back end. Secondary goals are to improve the convenience of the API for the use case of our front end, and make it easier for users and contributors to identify which back end code supports a given front end feature. We plan to do this by:
+
+- Removing Django models representing User DB objects (e.g., tables), replacing them with functions that query and act on those DB objects directly. Insofar as we need to enrich the metadata about User DB objects with Mathesar-managed metadata, we'll do that after gathering the relevant info from the User DB.
+- Changing our API to use a JSON-RPC (2.0) spec.
+
+## A Motivating Example: Getting the table info in a schema
+
+To get the table info in a schema,
```
[the remark this answers is not in the document]
Comment (kgodey) [reply docs/engineering/architecture/index.md:22]: Can you clarify in the spec as well?

File: docs/engineering/architecture/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kgodey) [reply docs/engineering/architecture/index.md:104]: Thanks, I assume you'll add this to the spec as well.

**Comment by mathemancer** [comment]
@seancolsen @kgodey This is ready for re-review. The main change is to add a section to `models.md` describing the models for beta. I've tried to add enough information that it should be possible to follow.

I've also made some more minor changes as per the review comments from @kgodey 

Still TODO is to add more info about the setup of the DB-layer functions (and the DB-layer schemata and eventual schemata in general).

**Review by seancolsen — APPROVED** [review]
Thanks for spelling the models out in more detail, @mathemancer. Generally this looks great! I have some critique, but it's mostly pretty minor at this point...

- There are some discrepancies between these models and Pavish's permissions [design](https://www.figma.com/file/S97Mma0hAy5Syh1w85RWoB/Permissions-UX?type=design&node-id=109-98&mode=design&t=VwLx6mfR8Lwcr4nI-0). For example:

    - Some roles are "default". Others are not.
    
        image

        The first role configured for a database because the default role and is subsequently used by any Mathesar admin users who don't have a role specifically configured. Each database needs to have one and only one default role. (This is my understanding, at least). This requirement would seem to warrant an additional boolean field on `UserDatabaseMap` (and potentially a CHECK constraint too).

    - It seems like the `UserDatabaseMap` model also needs a field to store this setting (or one like it):

        image

        While we _are_ still debating the best way to determine permissions for table and column metadata, as far as I can tell we're all agreed that _this_ is the place we want to store permissions _for explorations_.

        My understanding is that there will only be two options for this setting: one option that lets users run explorations but not modify their definition (i.e. read-only access to the exploration object); and another option that gives users read/write access to the exploration object. So this could potentially be a boolean field called `can_write_explorations` or `can_write_metadata`.

        With that field on, `UserDatabaseMap`, I'm not sure that we need the `DatabaseUIQueryRole` at all.

- `DatabaseServerCredential`

    - My intuition would be to keep the `db_server` field and actually to make it required. The use case of _"management of a common username and password across multiple servers"_ sounds intriguing but it also sounds like something we don't intend to support right away.

- This is a small nitpick, but I would consider using simpler nouns for the models. For example:

    - `DatabaseServer` → `Server`
    - `DatabaseServerCredential` → `Role` or `Credential`
    - `UserDatabaseMap` → maybe one of: `Authorization`, `Allowance`, `Placement`, `Admittance`, `Access`, `Grant`, `Mandate`, `Membership`, `Passport`, `Permission`
        - Although these suggested names might seem odd, I think we'd get used to them.
        - I find the current name especially challenging on this model because it doesn't contain any singular noun, and that makes it hard to talk about it within our team. Compare these two statements:

            > Do we want to allow one Mathesar user to have multiple authorizations for the same database?

            vs:
            
            > Do we want to allow one Mathesar user to have multiple user-database-map records for the same database?

- On a similar note, some of the column names are inconsistent. For example:

    - `UIQuery`.`database_id` vs `UserDatabaseMap`.`database`
    - `SharedQuery`.`credential` vs `UserDatabaseMap`.`role`

    For FK columns, I think it's nice to either: _always_ postfix with `_id`, or _never_ postfix. I don't have a strong opinion either way though. And the nouns should be consistent too.

**Review by pavish — APPROVED** [review]
I've already discussed this with @mathemancer and provided my approval over call. I see any change that might be required based on the UX we end up with for permissions can be easily implemented over these models.

**Comment by kgodey** [comment]
I read through the docs and didn't notice anything to flag. I am not doing a detailed critique because I think getting my head into all the details will take a while and probably not be worth it since I already discussed the product-level concerns with @mathemancer and it seems like the proposed architecture meets them.

Some questions / comments:

(1) @mathemancer, we're not installing any additional Mathesar schemas as part of this, right? I see we're adding new types, but I assume they'll be installed in the existing schemas?

(2) I don't think I agree with @seancolsen's proposed naming schema, "User Database Map records" seems much more easy to talk about to me than "Authorizations", "user database map" gives me everything I need to understand about the model in the name even if it takes a little bit longer to say.

I do agree with @seancolsen about consistency of `_id` and nouns.

(3) Since we're redoing the architecture, can we call them Explorations in the backend instead of UIQuery? It would be nice to align backend and frontend terminology where possible.

**Comment by mathemancer** [comment]
@kgodey @seancolsen @pavish This should probably have another round of re-review.

The main changes are:

- added a field to the `UserDBRoleMap` to note metadata permissions for a given user on a given DB.
- added a default credential fkey to the `Database` model. Any admin connecting to that Database will use the default credential unless configured otherwise in the `UserDBRoleMap`.
- Removed the no-longer-needed `UIQueryDatabaseRole` model. All needed functionality should be handled by logic around the `metadata_role` field in the `UserDBRoleMap`.

Minor changes are:

- all foreign keys are now named for the table they refer to, with no `_id` suffix.
- Changed "Database" to DB in most places for brevity.
- Removed an `oid` field from the `Database` model, since it wouldn't be used by beta, and might never be used.
- noted not-null constraints for all models.
- Changed `UIQuery` to `Exploration`

I also moved the old model notes to their own page, since I kept find/replacing the wrong things.

**Comment by mathemancer** [comment]
@kgodey To answer your remaining question:

> (1) @mathemancer, we're not installing any additional Mathesar schemas as part of this, right? I see we're adding new types, but I assume they'll be installed in the existing schemas?

Correct. However, at some point we should rename the current schemas for consistency. We should consider whether to go through that pain before or after beta.

**Comment by kgodey** [comment]
@mathemancer If we need to rename them, doing it while we're still in alpha seems preferable to later.

**Review by seancolsen — APPROVED** [review]
Brilliant. This is looking really good now!

I'm a visual person, so I made this as a reference for myself. Perhaps it will help others.

Screenshot_20240329_112252

It's not comprehensive, but it gives you a picture of the relationships. It's also saved here: https://dbdiagram.io/d/Mathesar-internal-6604b63cae072629ce23f20a

**Comment by mathemancer** [comment]
> Brilliant. This is looking really good now!
> 
> I'm a visual person, so I made this as a reference for myself. Perhaps it will help others.
> 
> image
> 
> It's not comprehensive, but it gives you a picture of the relationships. It's also saved here: https://dbdiagram.io/d/Mathesar-internal-6604b63cae072629ce23f20a

That's really great; thank you!

File: docs/engineering/architecture/permissions.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+- Privileges on DB objects are granted to DB roles, which may be accessible to some (or all) Mathesar users.
+- Anytime an RPC function requiring DB access runs, it runs using the connection defined via the `UserDatabaseMap`, with the associated privileges on that database.
+
+!!! question "UX Question"
+    Should the admin think in terms of groups of Mathesar users, or specifically in terms of connections when dealing with DB-level privileges?
+    
+!!! danger "Potential Confusion"
+    In case the admin or DBA want multiple Mathesar users to be able to modify various DB objects, three options are available:
+    - Give all relevant Mathesar users access to connect as the owning DB user.
+    - `GRANT` the owning DB role (ostensibly a user) to DB users connectable by the relevant Mathesar users.
```
Comment (pavish) [inline docs/engineering/architecture/permissions.md:39]: One of the UX thoughts was to have a configurable product setting which grants access to DB objects created via the Mathesar UI automatically to the roles belonging to all Mathesar users that have access to the database/schema.

Bringing this up so it's on your radar to mull over.

File: docs/engineering/architecture/permissions.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    - Have at least one DB superuser available for use with a connection, and give relevant Mathesar users access to that DB superuser.
+    
+## Metadata object privileges
+
+Examples of these are table properties like preview columns. Permissions for these are derived from permissions on relevant User DB objects. So, if a user wants to modify the preview template for a table, they would be checked against the permission to modify the underlying table structure. Thus, modifying this metadata will occur in a request to modify the associated table, and the privilege check will be whether the relevant DB user is allowed to `ALTER` that table.
```
Comment (pavish) [inline docs/engineering/architecture/permissions.md:44]: This needs to be updated based on our last discussion where we decided that all metadata permissions will now be identified using the `metadata\_role` column in `UserDBRoleMap `.

**Review by pavish — COMMENTED** [review]
@mathemancer I went through a finer comb this time, I think it looks great overall. I have a few questions/comments which require clarification.

File: docs/engineering/architecture/index.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    - `name` (multiple requests to user DB for each column)
+    - `type`
+    - `type_options` -- e.g., the precision specified for a `numeric` column
+
+All of this gets joined together, then sent back as a response from the API. 
+
+With the new architecture, to get the same info,
+
+1. The front end calls an RPC endpoint `/api/v0/rpc/`, calling a function `get_schema_table_details` with `database` and `schema` parameters. The database is identified by a Django id referring to a database model, and the schema is identified by an OID.
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/index.md:70]: The plan is to have everything at the same endpoint. This is typically how such things are organized. It also lets us use the inbuilt documentation, batching, and error-handling that come with the framework. It also aligns with the JSON-RPC 2.0 spec.

How strongly do you feel about this?

File: docs/engineering/architecture/models.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+## DBServerCredential
+
+| Column      | Type                     | Notes                             |
+|-------------|--------------------------|-----------------------------------|
+| id          | integer                  | pkey                              |
+| created\_at | timestamp with time zone |                                   |
+| updated\_at | timestamp with time zone |                                   |
+| username    | character varying        | not null                          |
+| password    | character varying        | encrypted; not null               |
+| db\_server  | integer                  | not null; references DBServer(id) |
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/models.md:64]: I wasn't planning for it to be unique. As you mention, we may have reason to have more than copy of such a pair.

File: docs/engineering/architecture/permissions.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+## User Database Map Model
+
+This map uses a `user_id, database` pair to look up the needed credentials, then provides a connection to the database using those credentials (if possible).
+
+## Adding Connections (backend perspective)
+
+Regardless of UI, the backend should receive a `POST` request to the new RPC endpoint defining a new connection. We'll use the same functions set up in [PR \#3348](https://github.com/mathesar-foundation/mathesar/pull/3348). The relevant info should be stored in the `Database`, `DatabaseServer`, and `DatabaseServerCredential` models. Also, if the user does not already have an entry defining their role on the given database, we could create such an entry automatically in the `UserDatabaseMap` model. This is optional, and doesn't really affect the architecture. Then, to let some other users access that connection, we will provide an RPC function that lets an admin set different users' credentials for the given database by creating or updating `UserDatabaseMap` resources. Note that this does _not_ directly modify anything to do with permissions on actual database objects (e.g., schemata or tables).
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/permissions.md:23]: "Connection" here is referring to the front end concept. It needs a username, password (or other auth), host, port, and database. That fully specifies a connection string. The info is actually stored over various models for reasons you indicated, e.g., adding a second connection to a given server doesn't require storing that server's host and port again, and adding a second connection to a given database on a given server doesn't require anything except another username and password.

(1) This is mostly the way I see that working, except we don't need to store an entry in the `UserDatabaseMap` for the admin (or any admin) user to use the default credential. I don't care too much about that distinction, though.

(2) I agree this is the way that should work.

File: docs/engineering/architecture/permissions.md (hunk 6 above)
Comment (mathemancer) [reply docs/engineering/architecture/permissions.md:39]: Good to know.

File: docs/engineering/architecture/permissions.md (hunk 7 above)
Comment (mathemancer) [reply docs/engineering/architecture/permissions.md:44]: Good catch, you're right

File: docs/engineering/architecture/models.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (pavish) [reply docs/engineering/architecture/models.md:64]: > configure the credentials when we could allow them to select from an already saved credential

Assuming that credentials are currently configured within the database page on the UX, this would break the scope of what it means to save a new connection since it'll break the boundaries of a database on the UI. It would also be a security concern since we don't want credentials configured within one database to be leaked to another.

>  I'd imagine we'd actually want to validate that the newly-entered password agrees with the stored password

It'll introduce new complexities such as figuring out which is the correct password. The stored password might be outdated. If we were to re-validate, do we now also update the old one? All these scenarios would complicate UX.

> To me, these are UX considerations that I'd be inclined to make based on the models
> I'd want to approach the models from a perspective of data integrity and normalization

I think the usecase should determine the modal structure and not the other way around. The purpose of this model is also to support the UX. Keeing it flexible would help us experiment with the UX more than having to restrict our UX affecting usability for little functional benefit.

I do think we can talk about this later, and merging the PR in.

**Review by pavish — APPROVED** [review]

File: docs/engineering/architecture/models.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (seancolsen) [reply docs/engineering/architecture/models.md:64]: @mathemancer, @pavish, and I discussed this on a call today. On the call I suggested two different directions to go, which I outline below as (B) and (C).

(In these diagrams I'm using ⭐ to indicate multi-column unique constraints.)

- **(A)** This is what the PR currently has (listed here for reference)

    Screenshot_20240329_130345


- **(B)** Databases and credentials are siblings within a DB server.

    image

    
    - Credentials would be shared across separate databases.
    - Updating a password in one place would apply to multiple databases.
    - The same is true for port numbers. To edit a port number, we'd need to make it clear to users That they'd be editing the port number of the `DBServer` object which could apply to many databases.
    - Databases, servers, and credentials could all be maintained separately from one another.
    - It would be possible to keep a credential stored within Mathesar while deleting all databases. Then you could connect to a new database using that same credential.

- **(C)** Each database owns a set of credentials.

    image

    - Credentials and servers are both scoped _underneath_ databases.
    - Since a database only has one server, the sever details are inlined into the database model.
    - Separate databases require separate credentials, and those credentials must be duplicated.
    - Any changes you make to server or credential settings within a database will only apply to that database.
    - Deleting a database from Mathesar would delete all the credentials associated with it.

The point I'm trying to make is that (B) and (C) are both sound approaches, while (A) is error-prone and hacky (in my opinion). With (A) the DB structure does not provide clarity on whether a credential should be shared or duplicated across multiple databases. That's the sort of ambiguity that I strive to minimize when designing database schemas. Ideally any database state should be valid application state. So if our application expects separate sets of credentials per-database, then we should enforce that logic at the data layer to avoid bugs. If we're _not sure_ if we want separate credential sets or not, then we should figure that out now instead of allowing for both structures to coexist in the data layer.

From a product perspective, I think it's worth considering (B) and (C) — and I think we ought to do this _before_ getting into anything with Figma (for the sake of time). To me, (C) seems simplest which is a strong point in its favor. Design and implementation would be much more straightforward than with (C). Approach (B) seems more powerful. But perhaps power users would be better served by additional features (post-beta) which allow for configuration code (e.g. in yaml). Honestly I don't think I have a strong opinion on (B) vs (C). I just would prefer that we avoid (A).

Apologies that I failed to notice the lack of unique constraint. I would have mentioned it sooner had I noticed.

[the remark this answers is not in the document]
Comment (seancolsen) [reply docs/engineering/architecture/models.md:64]: Another product-level thought on (B) vs (C)...

If a team of people is using Mathesar as an Airtable replacement, we need to give them something equivalent to an Airtable "Base" ­— i.e. a self-contained project that they create from scratch within Mathesar.

- If we want them to use _schemas_ to separate their projects, then (C) works fine. However we run into challenges with schema level permissions.
- If we want them to use _databases_ to separate their projects, then (B) seems slightly superior because the process of creating and removing databases could be more easily decoupled from managing credentials. Creating a new Airtable base is quite easy. Requiring users to enter a Postgres password for that step would be a usability impediment.

[the remark this answers is not in the document]
Comment (seancolsen) [reply docs/engineering/architecture/models.md:64]: BTW, Brent wanted the code for these diagrams. It's here:

https://dbdiagram.io/d/Mathesar-internal-AB-6604b63cae072629ce23f20a

https://dbdiagram.io/d/Mathesar-internal-C-6606dd2e37b7e33fd70f376a

[the remark this answers is not in the document]
Comment (mathemancer) [reply docs/engineering/architecture/models.md:64]: Taking options (B) and (C) as a given, I strongly prefer (B), since it's more flexible and more representative of the actual underlying structure of the objects involved. The only reason I still support (A) is that we already have a UI that needs that flexibility (or ambiguity), and (A) can easily be modified to (B) if we go that route.

I'm suspicious of (C), and any UI that depends on (C), since it introduces a misconception that we (and the user) will have to maintain, i.e., that a role on the PostgreSQL server is namespaced under a Database on that server. We have typically regretted creating simplifying abstractions that misrepresent the way PostgreSQL actually works on this project.

Regarding the concerns of @pavish w.r.t. organizational multitenancy, I think it's perfectly reasonable for that to be on a Mathesar installation level for now (rather than on a Database level).

[the remark this answers is not in the document]
Comment (seancolsen) [reply docs/engineering/architecture/models.md:64]: Cool. And if you both want to go with (A) I'll stop fussing about it, though I might have some more pointed questions and critique once we have designs in Figma.

[the remark this answers is not in the document]
Comment (kgodey) [reply docs/engineering/architecture/models.md:64]: This discussion seems like bikeshedding to me at this point, we can always add a unique constraint later and I don't see why we need to make a decision about whether or how to share credentials between databases right now. I'm going to make the executive decision that we're going with (A).

If the UX we end up with means that (A) can end up with invalid DB states, then we can implement (B). It's not like backend changes are verboten after this gets merged.

I do not think we should do (C) in any situation because it's not accurate to how credentials actually work and trying to build on an incorrect abstraction will probably cause us problems later.

[the remark this answers is not in the document]
Comment (kgodey) [reply docs/engineering/architecture/models.md:64]: If we were to make a decision about whether to share credentials now, I'd still argue for (A) anyway because implementing sharing credentials well seems like it would add scope to the beta (extra UX flows, security considerations, etc.) and I'm not sure that the feature of sharing credentials is valuable enough to add scope.

[the remark this answers is not in the document]
Comment (pavish) [reply docs/engineering/architecture/models.md:64]: As already stated, I'm in agreement with @mathemancer and @kgodey to go with (A).

**Comment by mathemancer** [comment]
Okay. I've updated the permissions page to match the new models. Also, it seems like we're at a stage of temporary agreement, or at least a truce. So, I'm going to merge this before anyone changes their mind.
